### PR TITLE
Throttle courier queues when the channel has rate limit redis key not…

### PR DESCRIPTION
**Pull request by norkans7**
… expired

### CODE REVIEW

**Comment by nicpottier** [comment]
This feels slightly odd in that it seems like we are basically using a redis key as a communication channel to move a channel into being throttled when we next try to pop a message off. Why are we doing it that way as opposed to moving the channel to the throttled list explicitly when we decide we are throttled? I guess the advantage of this method is we can set how long we throttle the channel for by tweaking the expiration but the downside is that every call to pop a message off a queue now has an extra lookup. I think I'd rather see our throttled logic expanded to just allow throttles more than a second if that's what is needed:
  https://github.com/nyaruka/courier/blob/main/queue/queue.go#L212
  
That could work by having both the queue and say a counter in the throttle list and that gets decremented in the dethrottler, moving anything that now has a count <= 0 to the active queue set.

File: handlers/whatsapp/whatsapp.go [hunk 1]
```diff
@@ -839,6 +843,16 @@ func sendWhatsAppMsg(msg courier.Msg, sendPath *url.URL, payload interface{}) (s
 	req, _ := http.NewRequest(http.MethodPost, sendPath.String(), bytes.NewReader(jsonBody))
 	req.Header = buildWhatsAppHeaders(msg.Channel())
 	rr, err := utils.MakeHTTPRequest(req)
+
+	if rr.StatusCode == 429 || rr.StatusCode == 503 {
+		rateLimitKey := fmt.Sprintf("rate_limit:%s", msg.Channel().UUID().String())
+		rc.Do("set", rateLimitKey, "engaged")
+		rc.Do("expire", rateLimitKey, 2)
```
Comment (rowanseymour) [inline handlers/whatsapp/whatsapp.go:850]: doesn't the 429 response tell us when we can retry ?

Comment (norkans7) [reply handlers/whatsapp/whatsapp.go:850]: No, we get 

```
{"meta":{"api_status":"stable","version":"2.35.4"},"errors":[{"code":1015,"title":"Too many requests","details":"Rate limiting engaged - Too Many Requests."}]}
```

Comment (rowanseymour) [reply handlers/whatsapp/whatsapp.go:850]: so we're backing down for 2 seconds but do we have any information if that's a reasonable back down?

Comment (norkans7) [reply handlers/whatsapp/whatsapp.go:850]: We are pausing the queue to send for 2 seconds and try again once we have the 429 status we pause for another 2 seconds

Comment (norkans7) [reply handlers/whatsapp/whatsapp.go:850]: https://developers.facebook.com/docs/whatsapp/api/rate-limits#capacity

The rate limit we are having is that we had more than 50 requests per second so stopping 2 seconds seems that will help reset the limit

Comment (rowanseymour) [reply handlers/whatsapp/whatsapp.go:850]: ok can you add a little comment to the code here explaining the rational for 2 and maybe mention that in future it should use header values

Comment (rowanseymour) [reply handlers/whatsapp/whatsapp.go:850]: and then let's get this out there!

**Review by rowanseymour — APPROVED** [review]

**Pull request merged**